src/components/ReanimatedDrawerLayout.tsx

```diff
@@ -0,0 +1,745 @@
+      drawerContainerStyle,
+      contentContainerStyle,
+      minSwipeDistance = defaultProps.minSwipeDistance,
+      edgeWidth = defaultProps.edgeWidth,
+      drawerLockMode = defaultProps.drawerLockMode,
+      useNativeAnimations = defaultProps.useNativeAnimations,
+    } = props;
+
+    const isFromLeft = drawerPosition === positions.Left;
+
+    const sideCorrection = isFromLeft ? 1 : -1;
+
+    const sideCorrectedDragX = useDerivedValue(
+      () => sideCorrection * dragX.value
+    );
+
+    // While closing the drawer when user starts gesture in the greyed out part of the window,
+    // we want the drawer to follow only once the finger reaches the edge of the drawer.
+    // See the diagram for reference. * = starting finger position, < = current finger position
+    // 1) +---------------+ 2) +---------------+ 3) +---------------+ 4) +---------------+
+    //    |XXXXXXXX|......|    |XXXXXXXX|......|    |XXXXXXXX|......|    |XXXXX|.........|
+    //    |XXXXXXXX|......|    |XXXXXXXX|......|    |XXXXXXXX|......|    |XXXXX|.........|
+    //    |XXXXXXXX|..<*..|    |XXXXXXXX|.<-*..|    |XXXXXXXX|<--*..|    |XXXXX|<-----*..|
+    //    |XXXXXXXX|......|    |XXXXXXXX|......|    |XXXXXXXX|......|    |XXXXX|.........|
+    //    |XXXXXXXX|......|    |XXXXXXXX|......|    |XXXXXXXX|......|    |XXXXX|.........|
+    //    +---------------+    +---------------+    +---------------+    +---------------+
+
+    const openValue = useDerivedValue(() => {
+      return interpolate(
+        drawerTranslation.value,
+        [0, drawerWidth],
+        [0, 1],
+        Extrapolation.CLAMP
+      );
+    });
+
+    const gestureOptions: {
+      useNativeDriver: boolean;
+      listener?: (event: { nativeEvent: { translationX: number } }) => void;
+    } = React.useMemo(
+      () => ({
+        useNativeDriver: useNativeAnimations,
+        listener: props.onDrawerSlide
+          ? (ev) => {
+              const translationX = Math.floor(
+                Math.abs(ev.nativeEvent.translationX)
+              );
+              const position = translationX / containerWidth;
+
+              props.onDrawerSlide?.(position);
+            }
+          : undefined,
+      }),
+      [containerWidth, props, useNativeAnimations]
+    );
+
+    if (props.onDrawerSlide) {
+      gestureOptions.listener = (ev) => {
+        const translationX = Math.floor(Math.abs(ev.nativeEvent.translationX));
+        const position = translationX / containerWidth;
+
+        props.onDrawerSlide?.(position);
+      };
+    }
+
+    const isDrawerOpen = useSharedValue(false);
+
+    const handleContainerLayout = ({ nativeEvent }: LayoutChangeEvent) => {
+      setContainerWidth(nativeEvent.layout.width);
+    };
+
+    const emitStateChanged = React.useCallback(
+      (newState: DrawerState, drawerWillShow: boolean) => {
+        'worklet';
+        props.onDrawerStateChanged &&
+          runOnJS(props.onDrawerStateChanged)?.(newState, drawerWillShow);
```

Comment:
Can we check whether `onDrawerStateChanged` is a worklet and call it on the UI thread if so?

Reply:
I've tried implementing this with a simple `runOnJS` drop-in replacement function:

```tsx
interface WorkletOrRegularFunction {
  // `unknown` and `never` won't work in this case
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (...args: any[]): unknown;
  __workletHash?: boolean;
}

const adaptCallback = (callback: WorkletOrRegularFunction) => {
  'worklet';
  return callback.__workletHash ? callback : runOnJS(callback);
};
```
```js
props.onDrawerSlide && adaptCallback(props.onDrawerSlide)(newOpenValue);
```

Where `adaptCallback` would replace `runOnJS` and then be run like `runOnJS`.

Unfortunately, i could not get such mechanism to not crash, executing a returned `runOnJS`-wrapped function seems to crash every single time with the following error:

`ERROR  ReanimatedError: Exception in HostFunction: [Reanimated] Attempted to extract from a HostObject that wasn't converted to a Shareable., js engine: reanimated`

I believe the crashes are related to one of these issues, but none of them exactly replicate their repro:
- https://github.com/software-mansion/react-native-reanimated/issues/5442
- https://github.com/software-mansion/react-native-reanimated/issues/5443

The shortest working alternative i could find looks like this:

```tsx
interface WorkletOrRegularFunction {
  // `unknown` and `never` won't work in this case
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (...args: any[]): unknown;
  __workletHash?: boolean;
}

const isWorklet = (cb?: WorkletOrRegularFunction) => {
  return !!cb?.__workletHash;
};
```
```js
// eslint-disable-next-line babel/no-unused-expressions
isWorklet(props.onDrawerOpen)
  ? props.onDrawerOpen?.()
  : props.onDrawerOpen && runOnJS(props.onDrawerOpen)();
```

Not sure if we want to use it, it's bulky but at least it's somewhat readable.

Reply:
Wouldn't something like `runCallback` instead of `adaptCallback` work? I.e. instead of doing something like this:
```ts
adaptCallback(props.onDrawerSlide)(newOpenValue);
```
do it like this:
```ts
runCallback(props.onDrawerSlide, newOpenValue);
```

Instead of returning the new function, simply call it inside.

Reply:
Unfortunately the second approach doesn't work either, `runCallback` would have to be a worklet running on UI, and `__workletHash` doesn't exist on UI. The `isWorkletFunction` function provided by `reanimated` doesn't work on UI either.

As a result, we'd have to check if a function is a worklet on JS, then read that data on UI, I tried doing that with the following method:

```ts
// this runs on JS, returns either a worklet or a remote function
const adaptCallback = (callback: WorkletOrRegularFunction) => {
  return callback.__workletHash ? callback : runOnJS(callback);
};

const onDrawerOpen = useMemo(
  () => adaptCallback(props.onDrawerOpen ?? noop),
  [props.onDrawerOpen]
);

onDrawerOpen();
```

Unfortunately, it looks like `runOnJS` stops working after being returned by another function, and behaves like a regular function which crashes on UI with the following error: 

```py
ReanimatedError: [Reanimated] Tried to synchronously call a non-worklet anonymous function on the UI thread.
```

Now I'm not really ok with creating a large boilerplate wrapper for each callback we want to use, since ideally this is something we want to apply to all the callbacks.
Here's how such boilerplate looks like:
<details>
<summary>
collapsed code
</summary>

```tsx
    const isDrawerStateWorklet = useMemo(
      () => isWorkletFunction(props.onDrawerStateChanged),
      [props.onDrawerStateChanged]
    );
    const onDrawerStateChanged = useCallback(
      (newState: DrawerState, drawerWillShow: boolean) => {
        'worklet';
        if (isDrawerStateWorklet) {
          props.onDrawerStateChanged?.(newState, drawerWillShow);
        } else {
          props.onDrawerStateChanged && runOnJS(props.onDrawerStateChanged);
        }
        // %% lint demands props, says props.onDrawerStateChanged is useless, likely bogus
      },
      [isDrawerStateWorklet, props]
    );

    const isDrawerSlideWorklet = useMemo(
      () => isWorkletFunction(props.onDrawerSlide),
      [props.onDrawerSlide]
    );
    const onDrawerSlide = useCallback(
      (position: number) => {
        'worklet';
        if (isDrawerSlideWorklet) {
          props.onDrawerSlide?.(position);
        } else {
          props.onDrawerSlide && runOnJS(props.onDrawerSlide);
        }
      },
      [isDrawerSlideWorklet, props]
    );

    const isDrawerOpenWorklet = useMemo(
      () => isWorkletFunction(props.onDrawerOpen),
      [props.onDrawerOpen]
    );
    const onDrawerOpen = useCallback(() => {
      'worklet';
      if (isDrawerOpenWorklet) {
        props.onDrawerOpen?.();
      } else {
        props.onDrawerOpen && runOnJS(props.onDrawerOpen);
      }
    }, [isDrawerOpenWorklet, props]);

    const isDrawerCloseWorklet = useMemo(
      () => isWorkletFunction(props.onDrawerClose),
      [props.onDrawerClose]
    );
    const onDrawerClose = useCallback(() => {
      'worklet';
      if (isDrawerCloseWorklet) {
        props.onDrawerClose?.();
      } else {
        props.onDrawerClose && runOnJS(props.onDrawerClose);
      }
    }, [isDrawerCloseWorklet, props]);
```

</details>

Another method (this one crashes):

<details>
<summary>
collapsed code
</summary>

```ts
const onDrawerOpen = useMemo(
  () =>
    isWorkletFunction(props.onDrawerOpen)
      ? props.onDrawerOpen
      : runOnJS(props.onDrawerOpen ?? noop),
  [props.onDrawerOpen]
);

const onDrawerClose = useMemo(
  () =>
    isWorkletFunction(props.onDrawerClose)
      ? props.onDrawerClose
      : runOnJS(props.onDrawerClose ?? noop),
  [props.onDrawerClose]
);
```
</details>
 
I'd prefer to introduce this feature in a separate PR, where it'll apply to both the `ReanimatedDrawerLayout` and `ReanimatedSwipeable`.